scipy/sparse/dok.py

```diff
@@ -474,9 +474,10 @@ def tocoo(self, copy=False):
             return coo_matrix(self.shape, dtype=self.dtype)
 
         idx_dtype = get_index_dtype(maxval=max(self.shape))
-        data = np.asarray(_list(self.values()), dtype=self.dtype)
-        indices = np.asarray(_list(self.keys()), dtype=idx_dtype).T
-        A = coo_matrix((data, indices), shape=self.shape, dtype=self.dtype)
+        data = np.fromiter(itervalues(self), dtype=self.dtype, count=self.nnz)
+        I = np.fromiter((i for i,_ in iterkeys(self)), dtype=idx_dtype, count=self.nnz)
+        J = np.fromiter((j for _,j in iterkeys(self)), dtype=idx_dtype, count=self.nnz)
```

Comment:
... except that this creates non-contiguous arrays, which generates a performance hit later on.